### ISSUE TITLE
Codist 6.7 Beta

### ISSUE THREAD

**Issue by wmjordan** [wmjordan/Codist#223]
## Download
[Codist 7983](https://github.com/wmjordan/Codist/files/10202220/Codist.zip)

[Codist 7977](https://github.com/wmjordan/Codist/files/10148187/Codist.zip)
[Codist 7963](https://github.com/wmjordan/Codist/files/10146076/Codist.zip)
[Codist 7933](https://github.com/wmjordan/Codist/files/10131372/Codist.zip)
[Codist 7924](https://github.com/wmjordan/Codist/files/10128618/Codist.zip)
[Codist 7923](https://github.com/wmjordan/Codist/files/10111404/Codist.zip)
[Codist 7912](https://github.com/wmjordan/Codist/files/10102511/Codist.zip)
[Codist 7909](https://github.com/wmjordan/Codist/files/10099696/Codist.zip)
[Codist 7906](https://github.com/wmjordan/Codist/files/10097479/Codist.zip)
[Codist 7856](https://github.com/wmjordan/Codist/files/10048460/Codist.zip)
[Codist 7831](https://github.com/wmjordan/Codist/files/10027237/Codist.zip)
[Codist 7819](https://github.com/wmjordan/Codist/files/10008317/Codist.zip)
[Codist 7805](https://github.com/wmjordan/Codist/files/9996534/Codist.zip)
[Codist 7801](https://github.com/wmjordan/Codist/files/9993853/Codist.zip)
[Codist 7797](https://github.com/wmjordan/Codist/files/9985580/Codist.zip)
[Codist 7794](https://github.com/wmjordan/Codist/files/9978590/Codist.zip)
[Codist 7787](https://github.com/wmjordan/Codist/files/9969799/Codist.zip)
[Codist 7760](https://github.com/wmjordan/Codist/files/9943758/Codist.zip)
[Codist 7749](https://github.com/wmjordan/Codist/files/9934138/Codist.zip)
[Codist 7736](https://github.com/wmjordan/Codist/files/9899272/Codist.zip)

## What's New
- General
  - [x] Added google to default search engines (7801, #226)
  - [x] Fixed improper configuration saving after initial load (7923, #229)
  - [x] Added a command to call refactorings provided by Codist (7977)
- Syntax Highlight
  - [x] Made compatible to VS 17.5 preview 1 (7787)
  - [x] Added highlight style for private members (7819)
- Resource Monitor
  - [x] Added option to display network adapter throughput per second (7736)
  - [x] Added hover to display top CPU usage processes to CPU monitor control (7760)
  - [x] Added option to customize task manager that launched after clicking monitor controls (7760)
- Super Quick Info
  - [x] Added implementation or declaration type info in interface list (#222, 7736, 7909, 7912, 7933)
  - [x] Changed behavior of the symbol signature icon in alternative style to display commands for generic type (7749)
  - [x] Merged some C# Quick Info options about type inheritance (7749)
  - [x] Displayed `GetEnumerator` method for `foreach` statement when hovering on `in` keyword (7760)
  - [x] Enabled Click & Go for C# label (7760)
  - [x] Added a new option to display all enum fields on C# Quick Info (7794, 7797, #225)
     <img width="305" alt="image" src="https://user-images.githubusercontent.com/12182730/201032767-a302555c-9fe0-49fc-a4d1-9aa24cb60538.png">
     <img width="214" alt="image" src="https://user-images.githubusercontent.com/12182730/201033327-8006eb5c-9cdc-4b4d-a90b-b6772e8bb507.png">
  - [x] Fixed symbol signature was not wrapped in C# Quick Info (7923)
  - [x] Added fallback for a bug in Roslyn which prevented C# Quick Info from applying Click & Go for signature value nodes (7963)
- Smart Bar
  - [x] Displayed symbol analysis commands for type argument for generic types with arity=1 (7749)
  - [x] Enabled symbol analysis commands for predefined types on C# Smart Bar (7749)
  - [x] Enabled reorder refactoring for argument type on C# Smart Bar (7749)
  - [x] Enabled Extract Method refactoring when debugging (7760)
  - [x] Optimized behavior of Join Lines command (7963)
  - [x] Added quite a few refactoring methods to C# Smart Bar (7787, 7805, 7819, 7831, 7856, 7906, 7923, 7933, 7963)
    <img width="393" alt="image" src="https://user-images.githubusercontent.com/12182730/200803115-161b42ef-060c-4158-aca8-5ed801967d06.png">
    <img width="356" alt="image" src="https://user-images.githubusercontent.com/12182730/200803455-c9dc816f-70de-4b0d-840c-1619799701e8.png">
  - [x] Added a new command to select all occurrences of current symbol in active document (7856)
    <img width="419" alt="image" src="https://user-images.githubusercontent.com/12182730/202883231-7ebbe897-d7bb-4a8f-8340-0a13674cf829.png">
  - [x] Added Copy without Indentation mode to the Copy command (7924)
    <img width="295" alt="image" src="https://user-images.githubusercontent.com/12182730/204947147-b23d6fcf-a30e-43e7-9e0c-ef84dc399753.png">
    For instance, select the following code, which is indented with 4 preceding tabs, then press Ctrl before clicking the copy button.
````C#
				if (line.Extent.IsEmpty) {
					spans.Add(line.Extent);
				}
````

Result:
````C#
if (line.Extent.IsEmpty) {
	spans.Add(line.Extent);
}
````
- Symbol Analysis Menu
  - [x] Changed style of symbol name in symbol analysis menu (7749)
- Navigation Bar
  - [x] Menu would pop up from wrong window when document windows were arranged with keyboard (7906, #228)
  - [x] Added unit test commands to C# NaviBar nodes (7963)
  - [x] Made Ctrl+Clicking on type nodes on C# NaviBar to go to type definition (7983, #233 )

**Comment by fitdev** [comment]
Just a quick feedback on 7794 (7795 - could not download because it said file was not found):

![Screenshot_20221110_165654](https://user-images.githubusercontent.com/8166847/201111188-76014a7b-1921-4ad0-913c-7914e746d6c5.png)

For some reason the constructors, as well as readonly struct types names in super quick infos are colorized as regular structs (not as readonly structs - I have 2 different styles for them).

**Comment by wmjordan** [comment]
@fitdev What's the VS version you are on?

**Comment by fitdev** [comment]
I see. It's not a bid deal, as I am still not sure if I want to distinguish between readonly and regular structs.

I am on VS 17.5 preview 1

**Comment by fitdev** [comment]
Thank you for letting me know. I was concerned about this actually. But so far I have not yet noticed anything yet, but then again I was not working for the last 10 days. Next week I will get into it and see if occurs. So far I had VS just sitting there with solution loaded and it did _not_ consume any more CPU than it did before, but perhaps it will once I start typing in the editor.
I do monitor CPU and other resources all the time through Process Hacker app, and will let you know if I have the same issue.

**Comment by wmjordan** [comment]
There is a new beta version with more code refactoring commands in C# Smart Bar.

**Comment by fitdev** [comment]
Just a quick report on another weird colorization:

![Screenshot_20221113_161438](https://user-images.githubusercontent.com/8166847/201523519-35c62201-13fb-4232-8ae6-22a22df0ad62.png)

All struct names should be Pinkish (the same color as for `ByteSize` since they are all readonly structs).

**Comment by wmjordan** [comment]
It is really a challenge to have the highlight work......as from time to time, VS changes its way...

**Comment by wmjordan** [comment]
> readonly structs

The problem is the `record struct`, which somehow has a higher precedence over the "C#: Read-only struct", no matter how I change their priority order. You have to change the color of the `record struct name`.

**Comment by fitdev** [comment]
Thanks for the suggestion. Yes, I will do that. Though it does raise a more general interesting question: how do you resolve such conflicts in general like when you have 2 modifiers applied which one takes precedence? I guess it is not something you can control in VS now. Maybe they will add some API in the future to control this.

**Comment by wmjordan** [comment]
> I guess it is not something you can control in VS now

You are right.
It is quite limited.
I somehow managed to find a way to control the merged styles. However, to fully support that, lots of work has to be done.

**Comment by fitdev** [comment]
A small issue with showing interface inheritance in superquick infos:

![Screenshot 2022-11-26 112549](https://user-images.githubusercontent.com/8166847/204079703-77ee7dd0-cd07-4389-82ab-390ecf942ffa.png)

For _classes_ it shows all interfaces as if they were explicitly declared on the class itself, instead of showing that they are actually coming from the base class as shown on the screenshot. Ideally, it should show interfaces with information where they are declared explicitly - base class, base of the base class, etc., or some interface name (it is common that some small interfaces are only declared on (inherited by) larger interfaces and not classes). So the tooltip should clearly show this, i.e. where a particular interface is "coming from".

**Comment by fitdev** [comment]
Another small thing: I noticed that *bold* style for _type declarations_ does not work (they are not bolded). I would like all type declarations like `public class Foo {}` to be *bold*, and there is an option for this in Codist, but it does not affect how it is displayed in the latest VS.

**Comment by wmjordan** [comment]
> So the tooltip should clearly show this, i.e. where a particular interface is "coming from".

I am sorry that I have to reconsider it.
The screenshot is showing information about a class. From the prospect of object-oriented programming, a class implements some interfaces--that is what we mostly care about when we use that class.

**Comment by wmjordan** [comment]
> I noticed that _bold_ style for _type declarations_ does not work (they are not bolded)

That's weird. I have never experienced this, even when the highlight was somewhat broken sometimes (#227) here.
The C# Declaration was by default Bold (see below) on Codist. Have you UNCHECKed the Bold box in another syntax style?
<img width="421" alt="image" src="https://user-images.githubusercontent.com/12182730/204087770-993c99f8-dc14-49f7-83f3-efbdc96a3ff9.png">

**Comment by fitdev** [comment]
> The screenshot is showing information about a class. From the prospect of object-oriented programming, a class implements some interfaces--that is what we mostly care about when we use that class.

I understand that but the way it currently appears on the screenshot is completely redundant: yes, it is true that `FsTreeEntryDir` class implements all those interface, but it does so _through its base class_ (i.e. `FsPathTreeDir`). So then why does the tooltip show all these interfaces as coming directly from `FsTreeEntryDir` (i.e. current class), instead of showing them as coming from `FsPathTreeDir` base classes (or its base classes as appropriate)?

In other words, the `: FsTreeEntryDir<This, Tree>`s at the end of most interfaces in the screenshot above are unnecessary, since the information they provide is not correct / unneeded (those interface come from the _base_ class(es), not the current class). And if, as you suggest, for classes you do not want to show the fact that certain interfaces are implemented / inherited through their base class(es), then why have this extra line ending at all, since they all show the same thing (`: FsTreeEntryDir<This, Tree>` in this example) _regardless of where the interface is actually coming from_?

**Comment by fitdev** [comment]
> Have you UNCHECKed the Bold box in another syntax style?

No, here's the screenshot:

![Screenshot 2022-11-26 155608](https://user-images.githubusercontent.com/8166847/204090107-a0afdb6c-3c1d-4a67-876f-17f48ce46888.png)

Ideally I would want everything underlined in red to be *bold*. But at the very least, the _name of the type with generic parameters_ should be bold. Or is this something not possible?

I am sorry I have upgraded VS to the latest preview and migrated my VS profile, and maybe that has screwed up certain things, so maybe I am asking for something which was never there in the first place.

**Comment by wmjordan** [comment]
From the above screenshot I think Codist is doing what is designed to do.

It won't have everything in the declaration node styled. It only applies the style to the name of the declared member,  demonstrated like the following screenshot, where the member declaration has an enlarged font size specified.

<img width="490" alt="image" src="https://user-images.githubusercontent.com/12182730/204126085-18f4f781-7cd4-4926-a761-4998bf81c138.png">

**Comment by wmjordan** [comment]
> I understand that but the way it currently appears on the screenshot is completely redundant: yes, it is true that `FsTreeEntryDir` class implements all those interface, but it does so _through its base class_ (i.e. `FsPathTreeDir`). So then why does the tooltip show all these interfaces as coming directly from `FsTreeEntryDir` (i.e. current class), instead of showing them as coming from `FsPathTreeDir` base classes (or its base classes as appropriate)?

How do you think it should be displayed like, an example?

**Comment by fitdev** [comment]
> I think Codist is doing what is designed to do.

It would be ok if it would stylize just the type name, but it does not make bold even that. The reason why ion the screenshot the type name looks a bit bold is because I use a slightly bolder font for all my (user types). But the bold (Declaration) style in the above screenshot was NOT applied to the type name. It's not a big deal, but if you could figure this out it would be nice.

> How do you think it should be displayed like, an example?

Well, in my example, in the tooltip where it has: `: FsTreeEntryDir<This, Tree>`, instead it should have `FsPathTreeDir<This, Tree>` because this is where those interfaces are coming from (i.e. coming from the base class: `FsTreeEntryDir<This, Tree>` inherits `FsPathTreeDir<This, Tree>`)! That is the name of the class / type that appears after the `:` when displaying a list of interfaces for a _class_ or _struct_ should be the type (either some base class) or the interface where that interface is defined.

Maybe the case of structs is a more complicated one, because a single interface can be implemented by more than one other interfaces, through which a particular struct will implement it, but at least it should work for classes: if the interface is NOT defined as being implemented by the class, then it is coming from its base class (or the base of the base, etc.), in which case, what should appear after `:` next to that interface is not the current class, but it's base class (or base of base - wherever the interface is actually implemented / "brought in").

**Comment by wmjordan** [comment]
> Well, in my example, in the tooltip where it has: `: FsTreeEntryDir<This, Tree>`, instead it should have `FsPathTreeDir<This, Tree>` because this is where those interfaces are coming from (i.e. coming from the base class: `FsTreeEntryDir<This, Tree>` inherits `FsPathTreeDir<This, Tree>`)!

Please try the new beta and see whether it is doing what is supposed to do.
If it doesn't, please show me the declaration code of relevant classes and interfaces.

**Comment by fitdev** [comment]
Thank you for the new version! It is a step in the right direction, however there are 2 issues still:

![Screenshot 2022-11-28 105020](https://user-images.githubusercontent.com/8166847/204222992-78251d80-68cc-4e12-a7d3-b5ebfba99908.png)

1. For some reason interfaces defined on the _immediate base class_ (`FsPathTreeDir<This, Tree>` in this example) are _still_ shown as if they were defined on the current class (i.e. on `FsEntryTreeDir<This, Tree>`). But they are not defined on the current class - they come from the base class. The only interface defined on the current class is `IFsEntryTreeDir<This>` and is shown correctly as such (no `:` after interface name because it comes from the current class directly).
2. If you compare this tooltip to a [previous tooltip](https://user-images.githubusercontent.com/8166847/204079703-77ee7dd0-cd07-4389-82ab-390ecf942ffa.png), some interfaces are _missing_. The missing interfaces are those that are not explicitly defined on any class (they come from other interfaces). I think for those there can be 2 possibilities:
- Use the same behavior as you do for superquickinfos for Interfaces, i.e. you show after `:` which _interface_ the interface being listed is "coming from"
- OR: Since we are showing quickinfo for a _class_, maybe show the base-most class name that implements the interface's containing interface.

Below is the breakdown of how it should work:

```cs

class A : B, IA { }

class B : C, IB { }

class C, IC { }

interface IA { }

interface IB : IExtra1 { }

interface IC : IExtra2 { }

interface IExtra1 { }

interface IExtra2 { }

```

The superquick info for `A`, interfaces section:

- `IA`
- `IB` : `B`
- `IC` : `C`
- `IExtra1` : `B` (because it comes from `IB` which comes from `B`) OR `IB` (because it comes from `IB`)
- `IExtra2` : `C` (because it comes from `IC` which comes from `C`) OR `IC` (because it comes from `IC`)

**Comment by wmjordan** [comment]
I reverted to the previous version, it was showing the correct info for `A`.

<img width="158" alt="image" src="https://user-images.githubusercontent.com/12182730/204234720-981c3c01-26b8-42f3-a47b-a635c44ce1b4.png">

Could you take the trouble of giving the definitions of your classes and interfaces in problem?

**Comment by wmjordan** [comment]
Could it be the problem like the following `IA` has already inherited from `IB`:
````C#
class A : B, IA { }

class B : C, IB { }

class C : IC { }

interface IA : IB { }

interface IB : IExtra1 { }

interface IC : IExtra2 { }

interface IExtra1 { }

interface IExtra2 { }
````

Then the Quick Info will display that `A` has implemented `IB` and `IExtra1`.

**Comment by fitdev** [comment]
> Could it be the problem like the following IA has already inherited from IB?

You are right. This is what was going on.

However to merely show `IB` : `A` is confusing because when you inspect declaration of `A` you will find `: IA` only and not `..., IB`. So I suggest that in the case of `A` (i.e. the _current_ class) we just show '`IB` : `IA`' because `IB` is implemented by `A` indirectly through `IA`, and there is no point in duplicating the fact that `A` consumes the interface directly (i.e. no point in showing '`IB` : `A`' because we already know that we are display superquickinfo for `A`). So better to use the same style for interfaces that are directly declared as implemented by a class (i.e. no ': Classname' after interface name, except that because an interface is not directly declared by a class but comes to it via intermediate interface, there should be ': IntermediateInterfaceName' instead).

So, elaborating on a more complex case, I suggest:

Given:

```cs
class A : B, IA { }

class B : C, IB, IExtra2, ICommon, IBOnly { }

class C : IC, ICommon, ICOnly { }

interface IA : IB { }

interface IB : IExtra1 { }

interface IBOnly : ICommonForBandC { }

interface ICOnly : ICommonForBandC { }

interface IC : IExtra2 { }

interface IExtra1 { }

interface IExtra2 { }

interface ICommon { }

interface ICommonForBandC { }
```

The SuperQuickInfo for `A` should show:

- `IA`
- `IB` : `IA` (it comes into `A` through `IA : IB`, but there is no point in showing just `IB` : `A`, and instead we are being more helpful and show through which interface `IB` comes into `A`, and in this case it comes through `IA`).
- `IC` : `C`
- `IExtra1` : `IB` (it comes directly into `A` through `A : IA : IB : IExtra1`, but we are being more helpful here and instead of showing `IExtra1` : `A` we show which immediate interface it is coming from)
- `IExtra2` : `C` (because it comes from `IC` which comes from `C`)
- `ICommon` : `B` (even though both `C` and `B` have this interface, we pick the most derived (closest) class.
- `IBOnly` : `B`
- `ICOnly` : `C`
- `ICommonForBandC` : `B` (just as with `ICommon` we pick the most derived class which contributes the interface (whether that class explicitly has the interface in its declaration or implements it through another interface or not) - in this case it's `B`

I think this covers all the cases including multiple interface inheritance.

I still think that in the future it would be helpful to provide even more information about the exact inheritance paths for interfaces, but for the moment this behavior would be great!

**Comment by wmjordan** [comment]
Thank you for helping out.
Please test the new version.

**Comment by fitdev** [comment]
Wow! You are so fast! Yes, this new version is great! Exactly what I was hoping for!

One small thing though, if possible, regarding the ordering of interfaces in the list, using the above example:

```cs
class A : B, IA { }

...

interface IA : IB, IAExtra { }

...

interface IAExtra { }

```

`IAExtra` is indirectly implemented by `A` through `IA`. Is it possible for all such cases to appear _right after_ `IA` in the list (i.e. after the interfaces that are directly present in the class' declaration) but _before_ interfaces that come from base classes?

**Comment by fitdev** [comment]
BTW, in my case I often deal with classes that are like 5 or 6 levels deep in inheritance hierarchy and have 40+ interfaces. So this feature is _very helpful_! Thank you!

**Comment by wmjordan** [comment]
> the ordering of interfaces in the list

I observed that before uploading the last beta. However, it will make the code more complicated at my first thought. 

Furthermore, I am not very sure whether the current implementation works well with base types as well.
Please email me if you find something wrong.

**Comment by wmjordan** [comment]
> the ordering of interfaces in the list

I observed that before uploading the last beta. However, it will make the code more complicated at my first thought. 

Furthermore, I am not very sure whether the current implementation works well with base types as well.
Please email me if you find something wrong.

**Comment by laicasaane** [comment]
@wmjordan I'm using the latest version:

![image](https://user-images.githubusercontent.com/1594982/205015152-b238d508-1abb-4777-bdad-fb31f3173402.png)

**Comment by wmjordan** [comment]
@laicasaane 
Please download the new beta.

**Comment by wmjordan** [comment]
I am about to release the final version in several days.
Please download and help test the latest beta.

**Comment by fitdev** [comment]
I noticed there are a lot of new refactorings in Codist. Is there a way to quickly invoke the refactoring menu with a single keyboard shortcut (without having to first show the Smart Bar and then use the mouse to select the refactoring menu)?

**Comment by wmjordan** [comment]
> Is there a way to quickly invoke the refactoring menu with a single keyboard shortcut

Hi, please download the new beta.
You can assign a keyboard shortcut for that function.
<img width="577" alt="image" src="https://user-images.githubusercontent.com/12182730/205483768-9a9fc923-379a-45cd-acbd-08eb4a394972.png">

**Comment by laicasaane** [comment]
Could you please add an option to delay showing quick info for some time? It often gets in my way of using multiple cursors. 😂

**Comment by wmjordan** [comment]
> delay showing quick info

It does not seem to be viable. No API to do so.

**Comment by laicasaane** [comment]
Thanks, I will try it now.
